lib/ansible/modules/database/mongodb/mongodb_replicaset.py

```diff
@@ -0,0 +1,407 @@
+
+    login_user = module.params['login_user']
+    login_password = module.params['login_password']
+    login_database = module.params['login_database']
+    login_host = module.params['login_host']
+    login_port = module.params['login_port']
+    replica_set = module.params['replica_set']
+    members = module.params['members']
+    arbiter_at_index = module.params['arbiter_at_index']
+    validate = module.params['validate']
+    ssl = module.params['ssl']
+    protocolVersion = module.params['protocolVersion']
+    chainingAllowed = module.params['chainingAllowed']
+    heartbeatTimeoutSecs = module.params['heartbeatTimeoutSecs']
+    electionTimeoutMillis = module.params['electionTimeoutMillis']
+
+    if validate:
+        if len(members) <= 2 or len(members) % 2 == 0:
+            raise UserWarning("MongoDB Replicaset validation failed. Invalid number of replicaset members.")
+        if arbiter_at_index is not None and len(members) - 1 > arbiter_at_index:
+            raise UserWarning("MongoDB Replicaset validation failed. Invalid arbiter index.")
+
+    result = dict(
+        changed=False,
+        replica_set=replica_set,
+    )
+
+    connection_params = dict(
+        host=login_host,
+        port=int(login_port),
+    )
+
+    if ssl:
+        connection_params["ssl"] = ssl
+        connection_params["ssl_cert_reqs"] = getattr(ssl_lib, module.params['ssl_cert_reqs'])
+
+    try:
+        client = MongoClient(**connection_params)
+    except Exception as e:
+        module.fail_json(msg='Unable to connect to database: %s' % to_native(e))
+
+    try:
+        check_compatibility(module, client)
+    except Exception as excep:
+        if "not authorized on" not in str(excep) and "there are no users authenticated" not in str(excep):
+            raise excep
+        if login_user is None or login_password is None:
+            raise excep
+        client.admin.authenticate(login_user, login_password, source=login_database)
+        check_compatibility(module, client)
+
+    if login_user is None and login_password is None:
+        mongocnf_creds = load_mongocnf()
+        if mongocnf_creds is not False:
+            login_user = mongocnf_creds['user']
+            login_password = mongocnf_creds['password']
+    elif login_password is None or login_user is None:
+        module.fail_json(msg="When supplying login arguments, both 'login_user' and 'login_password' must be provided")
+
+    try:
+        client['admin'].command('listDatabases', 1.0)  # if this throws an error we need to authenticate
+    except Exception as excep:
+        if "not authorized on" not in str(excep):
+            raise excep
+        if login_user is None or login_password is None:
+            raise excep
+        client.admin.authenticate(login_user, login_password, source=login_database)
+
+    if len(replica_set) == 0:
+        module.fail_json(msg="Parameter 'replica_set' must not be an empty string")
+
+    try:
```

Comment:
This try-except block should be reduce to the actual calls made.

Reply:
Should be sorted. My integration test are failing, early on during an apt install, not sure why yet. I have a bunch of playbooks using these module that have all passed. Output of one of these...

```
(py26env) [vagrant@AnsibleTest ~]$ ansible-playbook -l localhost playbooks/diet_2shard.yml 
 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the
implicit localhost does not match 'all'


PLAY [localhost] **************************************************************************

TASK [Gathering Facts] ********************************************************************
ok: [localhost]

TASK [Setup diet MongoDB Cluster] *********************************************************
changed: [localhost]

TASK [Wait a little bit] ******************************************************************
Pausing for 10 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
ok: [localhost]

TASK [Ensure replicaset rs0 exists] *******************************************************
changed: [localhost]

TASK [Ensure replicaset rs1 exists] *******************************************************
changed: [localhost]

TASK [Ensure shard rs0 exists] ************************************************************
changed: [localhost]

TASK [Wait a little bit] ******************************************************************
Pausing for 10 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
ok: [localhost]

TASK [Ensure shard rs1 exists] ************************************************************
changed: [localhost]

PLAY RECAP ********************************************************************************
localhost                  : ok=8    changed=5    unreachable=0    failed=0    skipped=0 
```